src/main/kotlin/com/exactpro/th2/rptdataprovider/producers/MessageProducer.kt

```diff
@@ -77,15 +90,39 @@ class MessageProducer(
                         parsedRawMessage[i]?.let {
                             Base64.getEncoder().encodeToString(it.body.toByteArray())
                         },
-                        processed?.get(i)?.get()?.metadata?.messageType ?: ""
+                        processed?.get(i)?.get()?.metadata?.messageType ?: "",
+                        attachedEvents?.get(i) ?: emptySet()
                     ).also { codecCache.put(it.messageId, it) }
-                }.associateBy { it.id }
+                }.associateBy { it.id },
+                needAttachEvents
             ).also {
                 codecCacheBatches.put(it.id.toString(), it)
             }
         }
     }
 
+    private suspend fun getAttachedEvents(
+        messages: MutableCollection<StoredMessage>
+    ): List<Set<String>> {
+        return coroutineScope {
+            messages.map { message ->
+                async {
+                    message.id.let {
+                        try {
+                            cradle.getEventIdsSuspend(it).map(Any::toString).toSet()
+                        } catch (e: Exception) {
+                            KotlinLogging.logger { }
```

Comment:
It is better to create a logger once (in companion object for example) and then use it